### PR TITLE
OAuth security and code quality overhaul

### PR DESCRIPTION
**Pull request by myssto**
Depends on #444

The OAuth implementation was written awhile ago now and was something I didn't fully understand. This PR contains some structure changes to the `OAuthHandler` and the way we read / write JWTs to push for more certainty in our authentication scheme

Changes:
- Abstracts the reading, writing, and validation of JWTs to a new service `IJwtService`
  - Now encodes the `User` or `Client` id into the `sub` claim instead of the `iss` claim (See [JWT Registered Claims](https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-token-claims#registered-claims). We were using the `iss` claim to store the id of the token owner. While this *technically* doesn't cause problems, its really really not great!)
  - Fixes refresh token validation bug
- All tokens now encode a `token-typ` claim into the payload that identifies the token as an access token or refresh token
  - Authentication step included to reject authentication for refresh tokens (In the current configuration, even though we do not encode a `User` or `Client`'s full privileges into the refresh token, they would still be able to access resources by providing their refresh token as a bearer token... not good!)
- Expands the `OtrClaims` mappings to more accurately adhere to the verbiage around JWTs
- Improve the `ClaimsPrincipleExtensions` to include more functionality
- Updates unit tests for `ClaimsPrincipleExtensions`

Config Change:
Adds field `Jwt.Issuer` to the configuration. This field is required, and should be the domain that the API is deployed to. For example on the staging branch of o!TR this would be `staging.otr.stagec.xyz`
```json
  "Jwt": {
    "Key": "",
    "Audience": "localhost:3000",
    "Issuer": "localhost:5075"
  },
```

### CODE REVIEW

**Comment by myssto** [comment]
Sorry in advance for size but should go down after #444 is merged. Also this would've been really hard to piece out, in fairness most of the size comes from renaming files.

File: API/Handlers/Implementations/OAuthHandler.cs [hunk 1]
```diff
@@ -1,40 +1,34 @@
...
     IOptions<AuthConfiguration> authConfiguration
     ) : IOAuthHandler
 {
-    private const int AccessDurationSeconds = 3600;
-    private const int RefreshDurationSeconds = 1_209_600;
-
-    public async Task<OAuthResponseDTO?> AuthorizeAsync(string osuAuthCode)
+    public async Task<AccessCredentialsDTO?> AuthorizeAsync(string osuAuthCode)
     {
         logger.LogDebug("Attempting authorization via osuAuthToken");
```
Comment (hburn7) [inline API/Handlers/Implementations/OAuthHandler.cs:33]: Let's change this to `osuAuthCode` while we're at it

File: API/Services/Implementations/JwtService.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+public class JwtService(
+    ILogger<JwtService> logger,
+    IOptions<JwtConfiguration> jwtConfiguration
+) : IJwtService
+{
+    private readonly JwtConfiguration _jwtConfiguration = jwtConfiguration.Value;
+
+    public int AccessDurationSeconds => 3600;
+    public int RefreshDurationSeconds => 1_209_600;
```
Comment (hburn7) [inline API/Services/Implementations/JwtService.cs:21]: For clarity either comment the total duration or use `(int) TimeSpan.FromX().TotalSeconds`

**Review by hburn7 — CHANGES_REQUESTED** [review]
Great work

**Comment by hburn7** [comment]
I don't think the audience and issuer here are any different in staging/prod. Unless you think the issuer should be `.../api`. Can you confirm?

**Comment by myssto** [comment]
> I don't think the audience and issuer here are any different in staging/prod. Unless you think the issuer should be `.../api`. Can you confirm?

The audience and issuer can be quite literally any string we want as long as the incoming tokens contain the same value that we set in the `ValidIssuer` and `ValidAudience` fields of the `TokenValidationParameters`. They are essentially just some added layers of protection. So to answer, yes we can set the issuer to be `otr.stagec.xyz/api` if we wanted.

File: API/Controllers/OAuthController.cs [hunk 3]
```diff
@@ -39,18 +38,24 @@ public async Task<IActionResult> AuthorizeAsync([FromQuery] string code)
...
-            ? Ok(result)
-            : Unauthorized();
+        if (!string.IsNullOrEmpty(result.ErrorDetail))
+        {
+            return BadRequest(result.ErrorDetail);
+        }
+
+        return result.Response is not null
+            ? Ok(result.Response)
+            : BadRequest("Unknown error");
```
[the remark this answers is not in the document]
Comment (myssto) [reply API/Controllers/OAuthController.cs:58]: What exactly do you mean by this? 520 is not a valid http code, unless you meant 500 which I can agree with. In any case if the `result.ErrorDetail` and `result.Response` are both null then clearly there is something wrong with the server side code. In that case I think a 500 error would be appropriate.

File: API/Services/Implementations/JwtService.cs (hunk 2 above)
Comment (myssto) [reply API/Services/Implementations/JwtService.cs:21]: These fields are documented in the interface `IJwtService`

```cs
    /// <summary>
    /// Lifetime of an access token in seconds
    /// </summary>
    public int AccessDurationSeconds { get; }

    /// <summary>
    /// Lifetime of a refresh token in seconds
    /// </summary>
    public int RefreshDurationSeconds { get; }
```

File: API/Controllers/OAuthController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hburn7) [reply API/Controllers/OAuthController.cs:58]: Okay turns out 520 is only for Cloudflare (I was googling what HTTP code is for 'unknown error') - I think 500 is fine.

File: API/Controllers/OAuthController.cs [hunk 4]
```diff
@@ -87,17 +80,23 @@ public async Task<IActionResult> CreateClientAsync()
...
```
[the remark this answers is not in the document]
Comment (hburn7) [reply API/Controllers/OAuthController.cs:100]: See correction comment

**Comment by hburn7** [comment]
> > I don't think the audience and issuer here are any different in staging/prod. Unless you think the issuer should be `.../api`. Can you confirm?
> 
> The audience and issuer can be quite literally any string we want as long as the incoming tokens contain the same value that we set in the `ValidIssuer` and `ValidAudience` fields of the `TokenValidationParameters`. They are essentially just some added layers of protection. So to answer, yes we can set the issuer to be `otr.stagec.xyz/api` if we wanted.

Sounds good.

**Review by hburn7 — APPROVED** [review]

**Pull request merged**